lib/IDE/Refactoring.cpp

```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
                                           EditConsumer).performChange();
 }
 
+static std::unique_ptr<llvm::SetVector<Expr*>> findConcatenatedExpressions(ResolvedRangeInfo Info) {
+  if (Info.Kind != RangeKind::SingleExpression && Info.Kind != RangeKind::PartOfExpression)
+    return nullptr;
+  Expr *E = Info.ContainedNodes[0].get<Expr*>();
+
+  struct StringInterpolationExprFinder: public SourceEntityWalker {
+    std::unique_ptr<llvm::SetVector<Expr*>> Bucket = llvm::make_unique<llvm::SetVector<Expr*>>();
+    bool IsValidInterpolation = true;
+    StringInterpolationExprFinder() {}
+
+    bool isConcatenationExpr(DeclRefExpr* Expr) {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD->getName().str() != "+") {
+        return false;
+      }
+      return true;
+    }
+
+    bool walkToExprPre(Expr *E) {
+      if (E->isImplicit())
+        return true;
+
+      auto ExprType = E->getType().getString();
+      //Only binary concatenation operators should exist in expression
+      if (E->getKind() == ExprKind::Binary) {
+        auto *BE = dyn_cast<BinaryExpr>(E);
+        auto *OperatorDeclRef = BE->getSemanticFn()->getMemberOperatorRef();
+        if (!isConcatenationExpr(OperatorDeclRef) || ExprType != "String") {
```

Comment:
Same here. We should use `ExprType->getNominalOrBoundGenericNominal() == Ctx.getStringDecl()` here instead of comparing strings.